Assets/Countly/Plugins/CountlySDK/Services/ViewCountlyService.cs

```diff
@@ -25,198 +25,882 @@ private class ViewData
+                string key = entry.Key;
+                ViewData vd = entry.Value;
+
+                if (vd != null && viewName.Equals(vd.ViewName)) {
+                    viewID = key;
                 }
+            }
 
-                if (name.Length > _configuration.MaxKeyLength) {
-                    Log.Verbose("[ViewCountlyService] RecordCloseViewAsync : Max allowed key length is " + _configuration.MaxKeyLength);
-                    name = name.Substring(0, _configuration.MaxKeyLength);
+            if (viewID == null) {
+                Log.Warning("[ViewCountlyService] StopViewWithNameInternal, No view entry found with the provided name :[" + viewName + "]");
+                return;
+            }
+
+            StopViewWithIDInternal(viewID, customViewSegmentation);
+        }
+
+        /// <summary>
+        /// Closes given views or all views.
+        /// </summary>
+        /// <param name="closeAllViews"></param>
+        /// <param name="customViewSegmentation"></param>
+        private void AutoCloseRequiredViews(bool closeAllViews, Dictionary<string, object> customViewSegmentation)
+        {
+            Log.Debug("[ViewCountlyService] AutoCloseRequiredViews");
+            List<string> viewsToRemove = new List<string>(1);
+
+            foreach (var entry in viewDataMap) {
+                ViewData vd = entry.Value;
+                if (closeAllViews || vd.IsAutoStoppedView) {
+                    viewsToRemove.Add(vd.ViewID);
                 }
+            }
+
+            if (viewsToRemove.Count > 0) {
+                Log.Debug("[ViewCountlyService] AutoCloseRequiredViews, about to close [" + viewsToRemove.Count + "] views");
+            }
+
+            _utils.RemoveReservedKeysFromSegmentation(customViewSegmentation, reservedSegmentationKeysViews, "[ViewCountlyService] AutoCloseRequiredViews, ", Log);
+
+            for (int i = 0; i < viewsToRemove.Count; i++) {
+                StopViewWithIDInternal(viewsToRemove[i], customViewSegmentation);
+            }
+        }
+
+        /// <summary>
+        /// Stops a view with the given ID if it was open.
+        /// </summary>
+        /// <param name="viewID">ID of the view</param>
+        /// <param name="viewSegmentation">view segmentation</param>
+        private void StopViewWithIDInternal(string viewID, Dictionary<string, object> customViewSegmentation)
+        {
+            if (_utils.IsNullEmptyOrWhitespace(viewID)) {
+                Log.Warning("[ViewCountlyService] StopViewWithIDInternal, Trying to record view with null or empty view ID, ignoring request");
+                return;
+            }
+
+            if (!viewDataMap.ContainsKey(viewID)) {
+                Log.Warning("[ViewCountlyService] StopViewWithIDInternal, there is no view with the provided view id to close");
+                return;
+            }
+
+            ViewData vd = viewDataMap[viewID];
+            if (vd == null) {
```

Comment:
isnt this check unnecessary. If map is already containing it, is there a possibility for it to be null?